### PR TITLE
fix(logging): bound secret patterns to the current line

### PR DESCRIPTION
**Pull request by mastertyko**
## Summary

Replace #2009 with a focused delta on current `main`. The original traceback leak is already fixed by #2053/#2054; what remains is pattern-bound hardening: secret patterns must not consume the next traceback line, unterminated JSON secrets must redact through end of line, and a glued Bearer `:` tail must not leak.

## Type of change

- [x] `fix:` — bug fix (no behavior change beyond the bug)

Linked issue: Related to #2009. Related to #2028 (deferred classes).

## Why this replaces the earlier PR

#2009 grew a grammar-aware authorization parser across many review rounds, then hard-conflicted with `_RedactingFormatterMixin` on `main`. Soju06 asked for a re-scoped rebase onto that mixin, or to close #2009 and fold the residual into #2028. This PR is that re-scope: one commit from current `main`, no formatter overrides, no `LogRecord` copy, WARNING+ keyed-secret gate unchanged.

## Known pre-existing gaps

Digest/AWS auth-param lists, quoted Python header keys, and whitespace-separated Bearer tails stay in #2028. They need a separate policy choice (fail-closed to end of line vs a credential grammar). This PR does not expand into that contract.

## OpenSpec

- [x] This PR includes / updates an OpenSpec change

Change directory: `openspec/changes/line-scope-log-secret-redaction/`

Delta on existing `proxy-runtime-observability` (baseline: `credential-safe-proxy-logging`). Not a new capability.

## Changes

- Apply keyed/bearer/authorization/JSON/Python-repr secret patterns per CR/LF line and reattach the original terminator bytes.
- Allow an unterminated JSON secret value to close at end of line.
- Include `:` in the Bearer token class so `Bearer abc.def:GLUEDTAIL, status=502` becomes `Bearer [REDACTED], status=502`.
- Leave same-line comma-less `Authorization: Basic … status=failed` truncation pinned.

## Simplicity

No new setting, README section, dashboard nav item, or default.

## Test plan

RED on `upstream/main` `dd28d7df` (`importlib` load of `app/core/runtime_logging.py`):

- `authorization=Basic X\nstatus=failed\napi_key=Y` → `'authorization=[REDACTED]'` (following lines swallowed)
- `payload={"token":"abc\nsafe diagnostic line` → token leaks
- `Bearer abc.def:GLUEDTAIL, status=502` → `Bearer [REDACTED]:GLUEDTAIL, status=502`

GREEN on this head:

```
.venv/bin/python -m pytest tests/unit/test_structured_logging.py -q
88 passed in 0.72s

.venv/bin/python -m pytest -q -ra --timeout=180 --timeout-method=thread tests/unit
7666 passed, 97 skipped, 12 warnings in 202.70s (0:03:22)

.venv/bin/ruff check app/core/runtime_logging.py tests/unit/test_structured_logging.py
All checks passed!

.venv/bin/ruff format --check app/core/runtime_logging.py tests/unit/test_structured_logging.py
2 files already formatted

uv run ty check app/core/runtime_logging.py tests/unit/test_structured_logging.py
All checks passed!

npx --yes @fission-ai/openspec@1.11.0 validate line-scope-log-secret-redaction --strict
Change 'line-scope-log-secret-redaction' is valid
```

`git diff --numstat` vs `dd28d7df`:

```
15      4       app/core/runtime_logging.py
58      0       openspec/changes/line-scope-log-secret-redaction/context.md
44      0       openspec/changes/line-scope-log-secret-redaction/design.md
38      0       openspec/changes/line-scope-log-secret-redaction/proposal.md
51      0       openspec/changes/line-scope-log-secret-redaction/specs/proxy-runtime-observability/spec.md
19      0       openspec/changes/line-scope-log-secret-redaction/tasks.md
72      0       tests/unit/test_structured_logging.py
```

Intentionally unrun locally: `cargo deny` / `make rust-audit` (cargo-deny is not installed here; covered by CI Rust workspace). Full `make ci` stopped at that missing tool after frontend, lint, typecheck, and rust-check had already passed.

## Checklist

- [x] Title is in Conventional Commits format.
- [x] Linked the related issue / discussion above.
- [x] Added or updated tests covering the change.
- [x] Ran focused ruff/ty/unit checks; full `make ci` blocked only on missing `cargo deny`.
- [x] `openspec validate line-scope-log-secret-redaction --strict` passes.
- [x] Simplicity gates reviewed.
- [x] CHANGELOG is **not** edited by hand.

## Summary by CodeRabbit

* **Bug Fixes**
  * Improved log redaction for bearer tokens containing colons and unterminated JSON-sensitive values, including trailing backslashes.
  * Prevented secrets on one log line from affecting subsequent lines, including traceback output.
  * Preserved original line breaks during redaction, including lone carriage returns.
  * Improved independent handling of Basic authentication tokens across multiline logs.
* **Tests**
  * Added coverage for multiline logs, JSON and bearer-token edge cases, line-ending preservation, and consistent repeated redaction.

### CODE REVIEW

**Comment by mastertyko** [comment]
@codex review

File: app/core/runtime_logging.py [hunk 1]
```diff
@@ -36,7 +37,7 @@
 _REDACTION_FAILED_PLACEHOLDER = "[REDACTED: log redaction failed]"
 _JSON_SENSITIVE_LOG_VALUE_PATTERN = re.compile(
     r'(?i)("(?:password|passwd|pwd|token|secret|api[_-]?key|authorization)"\s*:\s*")'
-    r'(?:\\.|[^"\\])*(")'
+    r'(?:\\.|[^"\\])*("|(?=\Z))'
```
[the remark this answers is not in the document]
Comment (mastertyko) [reply app/core/runtime_logging.py]: Fixed in f061532d. The JSON value loop now consumes a dangling `\` at end of line (`(?:\\(?=\Z))?`). New formatter regression `test_unterminated_json_secret_with_trailing_escape_redacts_through_end_of_line` keeps `QAJSONSECRET` out of text and JSON exception output and preserves the following diagnostic line.

File: app/core/runtime_logging.py [hunk 2]
```diff
@@ -98,6 +99,15 @@ def _redact_secret_patterns(text: str) -> str:
     return _PYTHON_REPR_SENSITIVE_LOG_VALUE_PATTERN.sub(_redact_python_repr_secret, redacted)
 
 
+def _redact_secret_patterns(text: str) -> str:
+    if "\n" not in text and "\r" not in text:
+        return _redact_secret_patterns_on_line(text)
```
[the remark this answers is not in the document]
Comment (mastertyko) [reply app/core/runtime_logging.py]: Fixed in f061532d. The all-level Basic prepass now uses the same `_map_log_lines` splitter as keyed patterns, so `Basic \\nSAFE` no longer treats `SAFE` as token material. `test_basic_token_prepass_does_not_swallow_following_line` pins it.

File: app/core/runtime_logging.py (hunk 1 above)
[the remark this answers is not in the document]
Comment (mastertyko) [reply app/core/runtime_logging.py]: Fixed in f061532d. Same trailing-escape closer as the Codex P1: `(?:\\(?=\Z))?` plus a formatter regression that asserts the token is gone and the next line remains.

File: openspec/changes/line-scope-log-secret-redaction/context.md [hunk 3]
```diff
@@ -0,0 +1,58 @@
...
+## Failure modes
+
+- Authorization without `,`/`&` still consumes the rest of **that** line,
+  including a same-line `status=failed`. That is existing pinned behavior.
+- DEBUG/INFO `exc_info=` still skips keyed patterns. Repository DEBUG sites
+  that log exceptions remain out of this change.
+
+## Example
+
+```
```
[the remark this answers is not in the document]
Comment (mastertyko) [reply openspec/changes/line-scope-log-secret-redaction/context.md]: Fixed in f061532d. Both example fences in `context.md` are now `text`.

File: tests/unit/test_structured_logging.py [hunk 4]
```diff
@@ -511,6 +511,78 @@ def test_authorization_midline_redaction_truncates_to_separator(text_formatter):
...
+        'payload={"token":"QAJSONSECRET\n'
+        "Bearer abc.def:GLUEDTAIL, status=502\n"
+    )
+
+    once = runtime_logging.redact_rendered_log_text(text)
+    twice = runtime_logging.redact_rendered_log_text(once)
+
+    assert once == twice
+    assert once.count("\n") == text.count("\n")
+    assert once.count("\r\n") == text.count("\r\n")
```
[the remark this answers is not in the document]
Comment (mastertyko) [reply tests/unit/test_structured_logging.py]: Fixed in f061532d. The test now compares the extracted `\\r\\n|\\n|\\r` sequence in order and includes a lone `\\r` line in the fixture.

**Comment by mastertyko** [comment]
@codex review

File: tests/unit/test_structured_logging.py [hunk 5]
```diff
@@ -511,6 +512,105 @@ def test_authorization_midline_redaction_truncates_to_separator(text_formatter):
...
+        raise RuntimeError('payload={"token":"' + token + "\\\nsafe diagnostic line")
+    except RuntimeError:
+        record = _record("request failed", exc_info=sys.exc_info())
+
+    output = _render(formatter, record)
+    rendered = json.loads(output)["exception"] if log_format == "json" else output
+
+    assert token not in rendered
+    assert '{"token":"[REDACTED]' in rendered
+    assert "safe diagnostic line" in rendered
```
[the remark this answers is not in the document]
Comment (mastertyko) [reply tests/unit/test_structured_logging.py]: Fixed in 0fd9a7c0. The trailing-escape JSON test now requires `{"token":"[REDACTED]\nsafe diagnostic line`, and the Basic prepass test asserts the full `Authorization: Basic \nSAFE diagnostic` byte string.

**Comment by mastertyko** [comment]
@codex review

**Comment by mastertyko** [comment]
Rerunning CI: fork cannot `gh run rerun --failed` (admin rights). Head `0fd9a7c0d952e49717ced15521ab4d444a8a4238` is unchanged; this close/reopen is the authorized self-service rerun for an unrelated integration-core flake (`test_codex_goal_restart_cannot_retire_owner_outside_api_key_scope` expected `hard_affinity_saturated`, got `upstream_request_timeout` after 75s). Same test passed on this PR at `db46322f` (https://github.com/Soju06/codex-lb/actions/runs/33934202225).
